### ISSUE TITLE
Is it process safe?

### ISSUE THREAD

**Issue by pyh007** [Delgan/loguru#929]
Is it process safe? For example, when uwsgi multi-process startup is used by my flask program, will there be log loss when the log file is rotated? I am worried that multiple processes will Rotate.

**Comment by Delgan** [comment]
Hi @pyh007,

Safety regarding multiprocessing is achievable by specifying enqueue=True to the sink and ensuring the logger is properly inherited by the child process. You can refer to [Compatibility with multiprocessing using the enqueue argument](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument) for more details.

However, using this approach with uwsgi can be challenging if the workers are not inherited as expected, and new processes are created from scratch. In such cases, the logs won't be lost during rotation because Loguru renames files instead of overwriting them. Yet, you might experience multiple files created at the same time during rotation when only one new file was expected.

I'm planning to implement a new method to ease interfacing with frameworks like Flask that rely on multiple spawned workers.

**Comment by pyh007** [comment]
When can you use the new method

**Comment by pyh007** [comment]
> 你好@pyh007,
> 
> `enqueue=True`通过指定接收器并确保`logger`子进程正确继承，可以实现多处理的安全性。您可以参阅[使用](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument)[参数的](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument)[兼容性`multiprocessing``enqueue`](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument)以获取更多详细信息。
> 
> `uwsgi`然而，如果工作人员没有按预期继承，并且新流程是从头开始创建的，那么使用这种方法可能会很困难。在这种情况下，日志在轮换期间不会丢失，因为 Loguru 会重命名文件而不是覆盖它们。然而，您可能会遇到在轮换过程中同时创建多个文件的情况，而实际上只需要一个新文件。
> 
> 我计划实现一种新方法来简化与 Flask 等依赖于多个生成工作线程的框架的交互。

When can you use the new method

**Comment by Delgan** [comment]
I can't provide any ETA for now, but I'm working on it.

**Comment by ayinger1** [comment]
any update on this?  i am noticing with multiprocessing the behavior explained above: logs not lost during rotation, but multiple archived files are created around the same time.  i was hoping that setting:   rotation="sunday at 00:00", context='spawn', enqueue=True   might help, but the issue remains.
(and i can't use 'fork' because i need it to be cross-platform, and i also use threading within each sub-process.)

**Comment by Delgan** [comment]
@ayinger1 The issue described in this PR is specific to Flask and others multi-workers web frameworks.

Does this match your use case? Generally speaking, Loguru _is_ process-safe. However, the current model does not integrate well with the web frameworks described earlier. Yet, if you are controlling `multiprocessing` usage in your program, you should be able to fully use Loguru without experiencing logging issues. 

One pitfall you need to watch out for is making sure that exactly one process is responsible for configuring the sinks. Sometimes, `logger.add()` is inadvertently called multiple times (by both the parent and child processes, for example). In such a case, multiple sinks write to the same file in parallel, and the problem only becomes apparent when the file is rotated.

See also: [Compatibility with `multiprocessing` using `enqueue` argument](https://loguru.readthedocs.io/en/latest/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument).

Feel free to open a new issue to discuss the problem you're facing more deeply.
